feat: redundant operations check

augustin-v opened the pull request:
Hello! This PR introduces a new lint rule that detects redundant operations including:
```
x + 0
x - 0
x / 1
x * 1
```
Note: no automatic code-fixing implemented yet

wawel37 commented on crates/cairo-lint-core/src/lints/redundant_op.rs:
```diff
@@ -0,0 +1,81 @@
+    diagnostics: &mut Vec<PluginDiagnostic>,
+) {
+    let func = function_trait_name_from_fn_id(db, &expr_func.function);
+    let op = func.split("::").last().unwrap_or("");
+
+    let is_redundant = match op {
+        "add" => is_zero(&expr_func.args[0], arenas) || is_zero(&expr_func.args[1], arenas),
+        "sub" => is_zero(&expr_func.args[1], arenas),
+        "mul" => is_one(&expr_func.args[0], arenas) || is_one(&expr_func.args[1], arenas),
+        "div" => is_one(&expr_func.args[1], arenas),
+        _ => false,
+    };
```
1. Don't just match with the operation name without the absolute path. It's better to match it to the predefined operation absolute paths in `crates/cairo-lint-core/src/lints/mod.rs`. See the `crates/cairo-lint-core/src/lints/erasing_op.rs` how it's matched there.

wawel37 commented on crates/cairo-lint-core/src/lints/redundant_op.rs:
```diff
@@ -0,0 +1,81 @@
+    }
+}
+
+fn is_zero(arg: &ExprFunctionCallArg, arenas: &Arenas) -> bool {
+    matches!(
+        arg,
+        ExprFunctionCallArg::Value(expr) if matches!(&arenas.exprs[*expr], Expr::Literal(val) if val.value == BigInt::from(0))
+    )
+}
```
Maybe it's better to move `crates/cairo-lint-core/src/lints/erasing_op.rs` `is_zero` function to `crates/cairo-lint-core/src/helper.rs`, so it can be used in multiple places, so we don't define 2 different functions for the same check.

wawel37 commented on crates/cairo-lint-core/tests/tests.rs:
```diff
@@ -453,3 +453,12 @@ test_file!(
     "int lt min one"
 );
+
+test_file!(
+    redundant_op,
+    redundant_op,
+    "multiplication by one",
+    "addition by zero",
+    "division by one",
+    "substraction by zero"
+);
```
Today we merged new tests model. Please adjust to the new changes. There's not a lot of them, but still

wawel37 commented on crates/cairo-lint-core/src/lints/redundant_op.rs:
```diff
@@ -0,0 +1,81 @@
+    )
+}
+
+fn is_one(arg: &ExprFunctionCallArg, arenas: &Arenas) -> bool {
+    matches!(
+        arg,
+        ExprFunctionCallArg::Value(expr) if matches!(&arenas.exprs[*expr], Expr::Literal(val) if val.value == BigInt::from(1))
+    )
+}
```
Same as with `is_zero`. I know it will be used only inside this lint rule, but this function seems to be rather general than this rule specific.

augustin-v commented:
Thank you for the review @wawel37, I rebased, updated the code to use absolute paths as instructed, and moved the `is_zero` and `is_one` functions in the `cairo-lint-core/src/helper.rs` as recommended. I've also updated the tests accordingly. Please let me know if there's anything else you need

wawel37 commented on crates/cairo-lint-core/src/helper.rs:
```diff
@@ -208,3 +209,55 @@ pub fn indent_snippet(input: &str, initial_indentation: usize) -> String {
+    matches!(
+        arg,
+        ExprFunctionCallArg::Value(expr)
+            if matches!(&arenas.exprs[*expr], Expr::Literal(val) if val.value == BigInt::from(0))
```
```suggestion
            if matches!(&arenas.exprs[*expr], Expr::Literal(val) if val.value == BigInt::ZERO)
```

wawel37 reviewed (approved):
LGTM, just please apply the suggested change to the `is_zero` function

wawel37 commented:
@augustin-v  Please make all your commits signed. Otherwise we won't be able to merge the PR

augustin-v commented:
> @augustin-v Please make all your commits signed. Otherwise we won't be able to merge the PR
Done
